compiler/src/iree/compiler/Codegen/Common/LinearizeMemRefs.cpp

```diff
@@ -0,0 +1,458 @@
+      return failure();
+    MemRefType newTypeOfSourceMemref = *maybeNewTypeOfSourceMemref;
+
+    SmallVector<Value> dynamicLinearizedSize;
+    if (!newTypeOfSourceMemref.hasStaticShape()) {
+      SmallVector<OpFoldResult> basis = getDimValues(
+          loc, rewriter, currentTypeOfSourceMemref, allocOp.getDynamicSizes());
+      SmallVector<Value> multiIndices(
+          basis.size(), rewriter.create<arith::ConstantIndexOp>(loc, 0));
+      multiIndices[0] = llvm::dyn_cast_if_present<Value>(basis[0]);
+      Value linearizedSizes = rewriter.create<affine::AffineLinearizeIndexOp>(
+          loc, multiIndices, basis, true);
+      dynamicLinearizedSize.push_back(linearizedSizes);
+    }
+
+    Value linearizedOp = rewriter.create<OpTy>(
+        loc, newTypeOfSourceMemref, dynamicLinearizedSize,
+        allocOp.getSymbolOperands(), allocOp.getAlignmentAttr());
+    SmallVector<int64_t, 2> indices(currentTypeOfSourceMemref.getRank());
+    std::iota(indices.begin(), indices.end(), 0);
+    Value delinearizedOp = rewriter.create<memref::ExpandShapeOp>(
+        loc, currentTypeOfSourceMemref, linearizedOp, ArrayRef({indices}),
+        allocOp.getMixedSizes());
+    rewriter.replaceOp(allocOp, delinearizedOp);
+    return success();
+  }
+};
+
+struct LinearizeMemrefLoad : public OpRewritePattern<memref::LoadOp> {
+  using OpRewritePattern<memref::LoadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(memref::LoadOp loadOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = loadOp->getLoc();
+    MemRefType currentTypeOfSourceMemref = loadOp.getMemRefType();
+    if (currentTypeOfSourceMemref.getRank() < 2 &&
+        loadOp.getIndices().size() < 2)
+      return success();
+    FailureOr<MemRefType> maybeNewTypeOfSourceMemref =
+        getLinearizedTypeFromSourceType(currentTypeOfSourceMemref);
+    if (failed(maybeNewTypeOfSourceMemref))
+      return failure();
+    MemRefType newTypeOfSourceMemref = *maybeNewTypeOfSourceMemref;
+
+    FailureOr<SmallVector<OpFoldResult>> basis =
+        getMixedOrigSize(loc, rewriter, loadOp.getMemref());
+    if (failed(basis))
+      return failure();
+    Value linearizedIndices = rewriter.create<affine::AffineLinearizeIndexOp>(
+        loc, loadOp.getIndices(), *basis, true);
+
+    SmallVector<Value> multiIndices(
+        (*basis).size(), rewriter.create<arith::ConstantIndexOp>(loc, 0));
+    multiIndices[0] = llvm::dyn_cast_if_present<Value>((*basis)[0]);
+    Value linearizedSizes = rewriter.create<affine::AffineLinearizeIndexOp>(
+        loc, multiIndices, *basis, true);
+    Value linearizedOperand =
+        linearizeOperand(loc, rewriter, loadOp.getMemref(),
+                         newTypeOfSourceMemref, linearizedSizes);
+    Value linearizedLoad = rewriter.create<memref::LoadOp>(
+        loc, linearizedOperand, linearizedIndices, loadOp.getNontemporal());
+
+    rewriter.replaceOp(loadOp, {linearizedLoad});
```

Comment:
nit: can use `rewrite.replaceOpWithNewOp<memref::LoadOp>` instead of the separate create + replace